File: 401. Binary Watch/C++/Solution.cpp
class Solution {
public:
    vector<string> readBinaryWatch(int num) {
        class Utils {
        public:
            void getCombination(int length, int numOnes, int base, vector<int> &pool) {
                if (numOnes == 0) {
                    pool.push_back(base);
                    return;
                }
                if (length == 0 || length < numOnes) {
                    return;
                }
                getCombination(length - 1, numOnes - 1, base + (1 << (length - 1)), pool);
                getCombination(length - 1, numOnes, base, pool);
            }
        };
        Utils utils;
        vector<string> result;
        for (int onesHour = 0; onesHour <= num; onesHour++) {
            int onesMinute = num - onesHour;
            vector<int> hours;
            vector<int> minutes;
            utils.getCombination(4, onesHour, 0, hours);
            utils.getCombination(6, onesMinute, 0, minutes);
            for (int hour : hours) {
                if (hour > 11) {
                    continue;
                }
                for (int minute : minutes) {
                    if (minute > 59) {
                        continue;
                    }
                    if (minute > 9) {
                        result.push_back(to_string(hour) + ":" + to_string(minute));
                    } else {
                        result.push_back(to_string(hour) + ":0" + to_string(minute));
                    }
                }
            }
        }
        return result;
    }
};

File: 401. Binary Watch/Python/Solution.py
class Solution(object):
    def readBinaryWatch(self, num):
        """
        :type num: int
        :rtype: List[str]
        """
        def get_combination(length, num_ones, base, pool):
            if num_ones == 0:
                pool.append(base)
                return
            if length == 0 or length < num_ones:
                return
            get_combination(length - 1, num_ones - 1, base + (1 << (length - 1)), pool)
            get_combination(length - 1, num_ones, base, pool)

        result = []
        for ones_hour in range(0, num + 1):
            ones_minute = num - ones_hour
            hours = []
            minutes = []
            get_combination(4, ones_hour, 0, hours)
            get_combination(6, ones_minute, 0, minutes)
            for hour in hours:
                if hour > 11:
                    continue
                for minute in minutes:
                    if minute > 59:
                        continue
                    result.append('%d:%02d' % (hour, minute))
        return result
